File: experiments/iso_ebnf_parser_v8.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import re
import json

# ---------------------------
# Lexer
# ---------------------------

class Token:
    def __init__(self, type_, value, line, col):
        self.type = type_
        self.value = value
        self.line = line
        self.col = col

    def __repr__(self):
        return f"Token({self.type}, {self.value}, {self.line}:{self.col})"

class Lexer:
    """Simple regex-based lexer for ISO EBNF with %RULES%/%TOKENS% blocks"""
    token_specification = [
        ('PERCENT', r'%'),
        ('EQUALS', r'='),
        ('SEMICOLON', r';'),
        ('PIPE', r'\|'),
        ('COMMA', r','),
        ('ASTERISK', r'\*'),
        ('PLUS', r'\+'),
        ('QUESTION', r'\?'),
        ('MINUS', r'-'),
        ('LBRACKET', r'\['),
        ('RBRACKET', r'\]'),
        ('LBRACE', r'\{'),
        ('RBRACE', r'\}'),
        ('LPAREN', r'\('),
        ('RPAREN', r'\)'),
        ('DOTDOT', r'\.\.'),
        ('HASH', r'#'),
        ('STRING', r'"([^"\\]|\\.)*"|\'([^\'\\]|\\.)*\''),
        ('IDENTIFIER', r'[A-Za-z_][A-Za-z0-9_]*'),
        ('INTEGER', r'\d+'),
        ('SPACE', r'[ \t\n\r]+'),
        ('COMMENT1', r'\(\*.*?\*\)'),
        ('COMMENT2', r'//[^\n]*'),
    ]
    def __init__(self, text):
        self.text = text
        self.tokens = []
        self.pos = 0
        self.line = 1
        self.col = 1
        self.regex = re.compile('|'.join('(?P<%s>%s)' % pair for pair in self.token_specification), re.DOTALL)
        self.tokenize()

    def tokenize(self):
        for mo in self.regex.finditer(self.text):
            kind = mo.lastgroup
            value = mo.group()
            if kind == 'SPACE' or kind.startswith('COMMENT'):
                self.update_pos(value)
                continue
            tok = Token(kind, value, self.line, self.col)
            self.tokens.append(tok)
            self.update_pos(value)

    def update_pos(self, text):
        lines = text.split('\n')
        if len(lines) > 1:
            self.line += len(lines) - 1
            self.col = len(lines[-1]) + 1
        else:
            self.col += len(text)

    def peek(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def next(self):
        tok = self.peek()
        if tok:
            self.pos += 1
        return tok

# ---------------------------
# Parser
# ---------------------------

class Parser:
    """Recursive-descent parser for ISO EBNF"""
    def __init__(self, lexer):
        self.lexer = lexer
        self.pos = 0
        self.tokens = lexer.tokens

    def current(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def advance(self):
        self.pos += 1

    def expect(self, type_):
        tok = self.current()
        if not tok or tok.type != type_:
            raise SyntaxError(f"Expected {type_} at pos {self.pos}, got {tok}")
        self.advance()
        return tok

    def parse_syntax(self):
        """Parse the entire syntax file with %RULES%/%TOKENS% blocks"""
        nodes = []
        mode = None
        while self.current():
            tok = self.current()
            if tok.type == 'PERCENT':
                directive = self.parse_directive()
                if directive == 'RULES':
                    mode = 'RULES'
                elif directive == 'TOKENS':
                    mode = 'TOKENS'
                else:
                    mode = None
            elif mode == 'RULES' and tok.type == 'IDENTIFIER':
                nodes.append(self.parse_rule())
            elif mode == 'TOKENS' and tok.type == 'IDENTIFIER':
                nodes.append(self.parse_token())
            else:
                self.advance()  # skip any unhandled token
        return nodes

    def parse_directive(self):
        self.expect('PERCENT')
        ident = self.expect('IDENTIFIER').value
        self.expect('PERCENT')
        return ident

    # ---------------------------
    # Rules Parsing
    # ---------------------------

    def parse_rule(self):
        rule_name = self.expect('IDENTIFIER').value
        self.expect('EQUALS')
        expr = self.parse_expression()
        self.expect('SEMICOLON')
        return {'rule': rule_name, 'expr': expr}

    def parse_expression(self):
        terms = [self.parse_term()]
        while self.current() and self.current().type == 'PIPE':
            self.expect('PIPE')
            terms.append(self.parse_term())
        if len(terms) == 1:
            return terms[0]
        return {'choice': terms}

    def parse_term(self):
        factors = [self.parse_factor()]
        while self.current() and self.current().type == 'COMMA':
            self.expect('COMMA')
            factors.append(self.parse_factor())
        if len(factors) == 1:
            return factors[0]
        return {'sequence': factors}

    def parse_factor(self):
        # First parse the primary element
        prim = self.parse_primary()

        # Then check if a quantifier follows
        quant = None
        tok = self.current()
        if tok and tok.type in ('ASTERISK','PLUS','QUESTION','LBRACE'):
            quant = self.parse_quantifier()

        if quant:
            return {'quantified': {'prim': prim, 'quant': quant}}
        return prim


    def parse_quantifier(self):
        tok = self.current()
        if tok.type == 'ASTERISK':
            self.advance()
            return {'min':0,'max':None}
        elif tok.type == 'PLUS':
            self.advance()
            return {'min':1,'max':None}
        elif tok.type == 'QUESTION':
            self.advance()
            return {'min':0,'max':1}
        elif tok.type == 'LBRACE':
            self.expect('LBRACE')
            min_val = int(self.expect('INTEGER').value)
            max_val = min_val
            if self.current() and self.current().type == 'COMMA':
                self.expect('COMMA')
                if self.current() and self.current().type == 'INTEGER':
                    max_val = int(self.expect('INTEGER').value)
                else:
                    max_val = None
            self.expect('RBRACE')
            return {'min': min_val,'max': max_val}
        else:
            return None

    def parse_primary(self):
        tok = self.current()
        if tok.type == 'IDENTIFIER':
            return {'identifier': self.expect('IDENTIFIER').value}
        elif tok.type == 'STRING':
            return {'terminal': self.expect('STRING').value}
        elif tok.type == 'LBRACKET':
            return self.parse_optional()
        elif tok.type == 'LBRACE':
            return self.parse_repeat()
        elif tok.type == 'LPAREN':
            return self.parse_group()
        else:
            raise SyntaxError(f"Unexpected primary {tok}")

    def parse_optional(self):
        self.expect('LBRACKET')
        expr = self.parse_expression()
        self.expect('RBRACKET')
        return {'optional': expr}

    def parse_repeat(self):
        self.expect('LBRACE')
        expr = self.parse_expression()
        self.expect('RBRACE')
        return {'repeat': expr}

    def parse_group(self):
        self.expect('LPAREN')
        expr = self.parse_expression()
        self.expect('RPAREN')
        return {'group': expr}

    # ---------------------------
    # Tokens Parsing
    # ---------------------------

    def parse_token(self):
        tok_name = self.expect('IDENTIFIER').value
        self.expect('EQUALS')
        tok_value = self.parse_token_value()
        self.expect('SEMICOLON')
        return {'token': tok_name, 'value': tok_value}

    def parse_token_value(self):
        tok = self.current()
        if tok.type in ('STRING','IDENTIFIER','INTEGER'):
            return self.expect(tok.type).value
        elif tok.type == 'HASH':
            self.expect('HASH')
            digits = self.expect('INTEGER').value
            return f'#{digits}'
        else:
            raise SyntaxError(f"Unexpected token value {tok}")

# ---------------------------
# Main
# ---------------------------

def main():
    if len(sys.argv) != 2:
        print("Usage: python3 iso_ebnf_parser_v8.py ISO_EBNF-file")
        sys.exit(1)
    filename = sys.argv[1]
    with open(filename,'r',encoding='utf-8') as f:
        text = f.read()
    lexer = Lexer(text)
    parser = Parser(lexer)
    grammar_json = parser.parse_syntax()
    print(json.dumps(grammar_json, indent=2))

if __name__ == "__main__":
    main()

File: experiments/iso_ebnf_parser_v8.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "iso_ebnf_lexer_v6.c"

/* -----------------------------
   AST definitions
------------------------------*/
typedef enum {
    NODE_RULE,
    NODE_SEQUENCE,
    NODE_CHOICE,
    NODE_PRIMARY,
    NODE_LITERAL,
    NODE_QUANTIFIER,
    NODE_CHARSET,
    NODE_CHAR_RANGE,
    NODE_SPECIAL_SEQ
} NodeType;

typedef struct ASTNode {
    NodeType type;
    char text[128];             // nome da regra ou literal
    struct ASTNode **children;  // sub-nós
    int child_count;
} ASTNode;

/* Criar novo nó */
ASTNode* create_node(NodeType type, const char *text) {
    ASTNode* node = (ASTNode*)malloc(sizeof(ASTNode));
    node->type = type;
    if(text) strncpy(node->text,text,127);
    else node->text[0]='\0';
    node->children=NULL;
    node->child_count=0;
    return node;
}

/* Adicionar filho */
void node_add_child(ASTNode *parent, ASTNode *child){
    parent->children = realloc(parent->children,sizeof(ASTNode*)*(parent->child_count+1));
    parent->children[parent->child_count++] = child;
}

/* -----------------------------
   Lexer globals
------------------------------*/
Lexer lexer;
Token current_token;

void next() { current_token = next_token(&lexer); }
int check(TokenType t) { return current_token.type == t; }
void expect(TokenType t) {
    if(check(t)) next();
    else{
        fprintf(stderr,"Parse error at %d:%d: expected token %d, got %d ('%s')\n",
            current_token.line,current_token.col,t,current_token.type,current_token.text);
        exit(1);
    }
}

/* -----------------------------
   Parser forward declarations
------------------------------*/
ASTNode* parse_syntax();
ASTNode* parse_rule();
ASTNode* parse_expression();
ASTNode* parse_term();
ASTNode* parse_factor();
ASTNode* parse_primary();
ASTNode* parse_optional();
ASTNode* parse_repeat();
ASTNode* parse_group();
ASTNode* parse_special_sequence();
ASTNode* parse_char_range();
ASTNode* parse_character_set();
ASTNode* parse_directive();

/* -----------------------------
   Parser implementation
------------------------------*/

/* syntax = { rule | directive | comment } */
ASTNode* parse_syntax() {
    ASTNode* root = create_node(NODE_SEQUENCE,"syntax");
    while(!check(TOK_EOF)){
        if(check(TOK_IDENTIFIER))
            node_add_child(root, parse_rule());
        else if(check(TOK_PERCENT))
            node_add_child(root, parse_directive());
        else next(); // ignora comentários
    }
    return root;
}

/* rule = identifier '=' expression ';' */
ASTNode* parse_rule() {
    char name[128];
    strcpy(name,current_token.text);
    next();
    expect(TOK_EQUALS);
    ASTNode* expr = parse_expression();
    expect(TOK_SEMICOLON);
    ASTNode* rule_node = create_node(NODE_RULE,name);
    node_add_child(rule_node,expr);
    return rule_node;
}

/* expression = term { '|' term } */
ASTNode* parse_expression() {
    ASTNode* term_node = parse_term();
    if(check(TOK_PIPE)){
        ASTNode* choice = create_node(NODE_CHOICE,"");
        node_add_child(choice,term_node);
        while(check(TOK_PIPE)){
            next();
            node_add_child(choice, parse_term());
        }
        return choice;
    }
    return term_node;
}

/* term = factor { ',' factor } */
ASTNode* parse_term() {
    ASTNode* factor_node = parse_factor();
    if(check(TOK_COMMA)){
        ASTNode* seq = create_node(NODE_SEQUENCE,"");
        node_add_child(seq,factor_node);
        while(check(TOK_COMMA)){
            next();
            node_add_child(seq,parse_factor());
        }
        return seq;
    }
    return factor_node;
}

/* factor = primary [ quantifier ] */
ASTNode* parse_factor() {
    ASTNode* primary = parse_primary();
    if(check(TOK_ASTERISK)||check(TOK_PLUS)||check(TOK_QUESTION)||check(TOK_LBRACE)){
        ASTNode* qnode = create_node(NODE_QUANTIFIER,current_token.text);
        next(); // consome quantificador simples ou abre brace
        if(primary) node_add_child(qnode,primary);
        if(check(TOK_LBRACE)){
            // parse {n} or {n,m}
            next();
            if(check(TOK_INTEGER)) next();
            if(check(TOK_COMMA)){
                next();
                if(check(TOK_INTEGER)) next();
            }
            expect(TOK_RBRACE);
        }
        return qnode;
    }
    return primary;
}

/* primary = identifier | terminal | optional | repeat | group | special_sequence | char_range | character_set */
ASTNode* parse_primary() {
    if(check(TOK_IDENTIFIER)||check(TOK_STRING)||check(TOK_HEX_CHAR)){
        ASTNode* node = create_node(NODE_LITERAL,current_token.text);
        next();
        return node;
    } else if(check(TOK_LBRACKET)) return parse_optional();
    else if(check(TOK_LBRACE)) return parse_repeat();
    else if(check(TOK_LPAREN)) return parse_group();
    else if(check(TOK_SPECIAL_SEQ)) return parse_special_sequence();
    else if(check(TOK_HASH)) return parse_char_range();
    else{
        fprintf(stderr,"Parse error at %d:%d unexpected token '%s'\n",
            current_token.line,current_token.col,current_token.text);
        exit(1);
    }
    return NULL;
}

/* optional = '[' expression ']' */
ASTNode* parse_optional(){
    expect(TOK_LBRACKET);
    ASTNode* node = create_node(NODE_SEQUENCE,"optional");
    node_add_child(node,parse_expression());
    expect(TOK_RBRACKET);
    return node;
}

/* repeat = '{' expression '}' */
ASTNode* parse_repeat(){
    expect(TOK_LBRACE);
    ASTNode* node = create_node(NODE_SEQUENCE,"repeat");
    node_add_child(node,parse_expression());
    expect(TOK_RBRACE);
    return node;
}

/* group = '(' expression ')' */
ASTNode* parse_group(){
    expect(TOK_LPAREN);
    ASTNode* node = create_node(NODE_SEQUENCE,"group");
    node_add_child(node,parse_expression());
    expect(TOK_RPAREN);
    return node;
}

/* special_sequence = '? ... ?' */
ASTNode* parse_special_sequence(){
    ASTNode* node = create_node(NODE_SPECIAL_SEQ,current_token.text);
    next();
    return node;
}

/* char_range = '#' hex_char '..' '#' hex_char */
ASTNode* parse_char_range(){
    expect(TOK_HASH);
    expect(TOK_HEX_CHAR);
    expect(TOK_DOTDOT);
    expect(TOK_HASH);
    expect(TOK_HEX_CHAR);
    return create_node(NODE_CHAR_RANGE,"#..#");
}

/* character_set = '[' character_element { '-' character_element } ']' */
ASTNode* parse_character_set(){
    expect(TOK_LBRACKET);
    ASTNode* node = create_node(NODE_CHARSET,"charset");
    while(!check(TOK_RBRACKET)){
        if(check(TOK_IDENTIFIER)||check(TOK_STRING)||check(TOK_HEX_CHAR)||check(TOK_SPECIAL_SEQ)){
            node_add_child(node,create_node(NODE_LITERAL,current_token.text));
            next();
        }
        if(check(TOK_MINUS)){
            next(); // consome -
        }
    }
    expect(TOK_RBRACKET);
    return node;
}

/* directive = '%' identifier [ '(' literal ')' ] '%' */
ASTNode* parse_directive(){
    expect(TOK_PERCENT);
    char name[128];
    strcpy(name,current_token.text);
    expect(TOK_IDENTIFIER);
    ASTNode* node = create_node(NODE_RULE,name);
    if(check(TOK_LPAREN)){
        next();
        if(!check(TOK_RPAREN)) node_add_child(node,parse_primary());
        expect(TOK_RPAREN);
    }
    expect(TOK_PERCENT);
    return node;
}

/* -----------------------------
   Print JSON recursively
------------------------------*/
void print_node_json(ASTNode* node,int indent){
    for(int i=0;i<indent;i++) printf(" ");
    switch(node->type){
        case NODE_RULE: printf("{\"type\":\"rule\",\"name\":\"%s\",",node->text); break;
        case NODE_SEQUENCE: printf("{\"type\":\"sequence\","); break;
        case NODE_CHOICE: printf("{\"type\":\"choice\","); break;
        case NODE_LITERAL: printf("{\"type\":\"literal\",\"value\":\"%s\"}",node->text); return;
        case NODE_QUANTIFIER: printf("{\"type\":\"quantifier\",\"symbol\":\"%s\",",node->text); break;
        case NODE_CHARSET: printf("{\"type\":\"charset\","); break;
        case NODE_CHAR_RANGE: printf("{\"type\":\"char_range\"}"); return;
        case NODE_SPECIAL_SEQ: printf("{\"type\":\"special_seq\",\"value\":\"%s\"}",node->text); return;
    }
    if(node->child_count>0){
        printf("\"children\":[\n");
        for(int i=0;i<node->child_count;i++){
            print_node_json(node->children[i],indent+2);
            if(i+1<node->child_count) printf(",\n");
        }
        printf("]}");
    } else printf("}");
}

/* -----------------------------
   Main
------------------------------*/
int main(int argc,char **argv){
    if(argc<2){ printf("Usage: %s file.ebnf\n",argv[0]); return 1; }
    FILE *f=fopen(argv[1],"r");
    if(!f){ perror("fopen"); return 1; }
    fseek(f,0,SEEK_END);
    long len=ftell(f); fseek(f,0,SEEK_SET);
    char *input=malloc(len+1);
    fread(input,1,len,f); input[len]='\0'; fclose(f);

    lexer.input=input; lexer.pos=0; lexer.line=1; lexer.col=1;
    next();

    ASTNode* ast = parse_syntax();

    printf("{\n\"ast\": ");
    print_node_json(ast,2);
    printf("\n}\n");

    free(input);
    return 0;
}
